core/templates/pages/exploration-player-page/services/exploration-engine.service.spec.ts

```diff
@@ -0,0 +1,910 @@
+        answer, textInputService, submitAnswerSuccessCb);
+
+      expect(alertsServiceSpy)
+        .toHaveBeenCalledWith('Question name should not be empty.');
+    });
+  });
+
+  it('should check whether we can ask learner for answer ' +
+    'details', fakeAsync(() => {
+    let initSuccessCb = jasmine.createSpy('success');
+
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+    spyOn(explorationFeaturesBackendApiService, 'fetchExplorationFeaturesAsync')
+      .and.returnValue(Promise.resolve(explorationFeatures));
+
+    // Here default value is set to false.
+    expect(explorationEngineService.getAlwaysAskLearnerForAnswerDetails())
+      .toBe(false);
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+    tick();
+
+    const answerDetails = (
+      explorationEngineService.getAlwaysAskLearnerForAnswerDetails());
+    expect(answerDetails).toBe(true);
+  }));
+
+  it('should return default exploration Id', () => {
+    // Please note that default exploration Id is 'test_id'.
+    // This is being initialized in the constructor.
+
+    const explorationId = explorationEngineService.getExplorationId();
+    expect(explorationId).toBe('test_id');
+  });
+
+  it('should return exploration title', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    expect(() => {
+      explorationEngineService.getExplorationTitle();
+    }).toThrowError('Cannot read property \'title\' of undefined');
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    const explorationTitle = explorationEngineService.getExplorationTitle();
+    expect(explorationTitle).toBe('My Exploration Title');
+  });
+
+  it('should return exploration version', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    // Here 1 is default value, this is being initialized in the constructor.
+    expect(explorationEngineService.getExplorationVersion()).toBe(1);
+
+    explorationEngineService.init(
+      explorationDict, 2, null, true, ['en'], initSuccessCb);
+
+    const explorationVersion = explorationEngineService.getExplorationVersion();
+    expect(explorationVersion).toBe(2);
+  });
+
+  it('should return author recommended exploration Id\'s', () => {
+    let initSuccessCb = jasmine.createSpy('success');
+
+    spyOn(contextService, 'isInExplorationEditorPage').and.returnValue(false);
+
+    expect(() => {
+      explorationEngineService.getAuthorRecommendedExpIds();
+    }).toThrowError(
+      'Cannot read property \'getAuthorRecommendedExpIds\' of undefined');
+
+    explorationEngineService.init(
+      explorationDict, 1, null, true, ['en'], initSuccessCb);
+
+    explorationEngineService.currentStateName = 'End';
```

Comment:
what happens if state name is not End?

Reply:
The state should be final/terminal in order to get the author recommended explorations.
Added prechecks and comments.